handler.go

```diff
@@ -0,0 +1,343 @@
+	transport  transport.Transport
+	resolved   resolvedProtocolEncoding
+	serializer encoding.Serializer
+	body       io.Reader
+	headers    map[string]string
+}
+
+func (r requestHandler) handle() {
+	if isStreamingMethod(r.serializer) {
+		r.handleStreamRequest()
+		return
+	}
+
+	r.handleUnaryRequest()
+}
+
+// handleUnaryRequest launches initial unary request and unary benchmark
+func (r requestHandler) handleUnaryRequest() {
+	reqInput, err := ioutil.ReadAll(r.body)
+	if err != nil {
+		r.out.Fatalf("Failed while reading the input: %v\n", err)
+	}
+
+	req, err := r.serializer.Request(reqInput)
+	if err != nil {
+		r.out.Fatalf("Failed while serializing the input: %v\n", err)
+	}
+
+	req, err = prepareRequest(req, r.headers, r.opts)
+	if err != nil {
+		r.out.Fatalf("Failed while preparing the request: %v\n", err)
+	}
+
+	// Decides if warm requests must be dispatched before benchmark.
+	if r.shouldMakeInitialRequest() {
+		makeInitialRequest(r.out, r.transport, r.serializer, req)
+	}
+
+	runBenchmark(r.out, r.logger, r.opts, r.resolved, benchmarkMethod{
+		serializer: r.serializer,
+		req:        req,
+	})
+}
+
+// handleStreamRequest launches initial stream request and stream benchmark
+func (r requestHandler) handleStreamRequest() {
+	streamSerializer, ok := r.serializer.(encoding.StreamSerializer)
+	if !ok {
+		r.out.Fatalf("Serializer does not support streaming: %v\n", r.serializer.Encoding())
+	}
+
+	streamReq, streamMsgReader, err := streamSerializer.StreamRequest(r.body)
+	if err != nil {
+		r.out.Fatalf("Failed to create streaming request: %v\n", err)
+	}
+
+	streamReq.Request, err = prepareRequest(streamReq.Request, r.headers, r.opts)
+	if err != nil {
+		r.out.Fatalf("Failed while preparing the request: %v\n", err)
+	}
+
+	nextBodyFn := streamRequestSupplier(streamMsgReader)
+
+	if r.shouldMakeInitialRequest() {
+		if err = makeStreamRequest(r.transport, streamReq, r.serializer, nextBodyFn, r.responseHandler); err != nil {
+			r.out.Fatalf("%v\n", err)
+		}
+	}
+
+	// TODO: handle streaming benchmark.
+}
+
+// shouldMakeInitialRequest returns true if initial request must be made
+func (r requestHandler) shouldMakeInitialRequest() bool {
+	return !(r.opts.BOpts.enabled() && r.opts.BOpts.WarmupRequests == 0)
+}
+
+// responseHandler validates the response bytes and prints indented JSON body
+func (r requestHandler) responseHandler(body []byte) error {
+	res, err := r.serializer.Response(&transport.Response{Body: body})
+	if err != nil {
+		return fmt.Errorf("Failed while serializing stream response: %v", err)
+	}
+
+	bs, err := json.MarshalIndent(res, "", "  ")
+	if err != nil {
+		return fmt.Errorf("Failed to convert map to JSON: %v\nMap: %+v", err, res)
+	}
+
+	r.out.Printf("%s\n\n", bs)
+	return nil
+}
+
+// streamRequestSupplier returns a stream request supplier function which uses
+// provided stream request reader
+func streamRequestSupplier(streamMsgReader encoding.StreamRequestReader) streamRequestSupplierFn {
```

Comment:
Should we add a receiver to this function. I agree this is not needed yet, but it might be in the future (with option-interval). It will also make the code more consistent with the response handler